<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Unpaired Multiomics &mdash; scBIOT 1.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=def9ab29"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            scBIOT
          </a>
              <div class="version">
                1.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Main content</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing.html">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../ot.html">Optimal transport: <code class="docutils literal notranslate"><span class="pre">ot</span></code></a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scBIOT</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Unpaired Multiomics</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/unpaired_multiomics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="unpaired-multiomics">
<h1>Unpaired Multiomics<a class="headerlink" href="#unpaired-multiomics" title="Permalink to this heading">ÔÉÅ</a></h1>
<p><a class="reference download internal" download="" href="../_downloads/572886e9e08a3c23b185e3e8d6e871a8/unpaired_Yao-2021_integration.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Open</span> <span class="pre">notebook</span></code></a> to match modalities with optimal transport when cells are unpaired.</p>
<p>Preprocess ATAC, harmonize gene names, and build an aligned co-embedding:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scbiot.ot.coembedding</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">build_aligned_coembedding</span><span class="p">,</span>
    <span class="n">harmonize_gene_names</span><span class="p">,</span>
    <span class="n">label_transfer_shared_pca</span><span class="p">,</span>
    <span class="n">preprocess_atac</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">atac_prep</span> <span class="o">=</span> <span class="n">preprocess_atac</span><span class="p">(</span>
    <span class="n">adata_atac</span><span class="p">,</span>
    <span class="n">gtf_file</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/inputs/gencode.vM25.chr_patch_hapl_scaff.annotation.gtf.gz&quot;</span><span class="p">,</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_ga</span> <span class="o">=</span> <span class="n">atac_prep</span><span class="o">.</span><span class="n">ga</span>

<span class="n">harmonize_gene_names</span><span class="p">(</span>
    <span class="n">adata_gex</span><span class="p">,</span>
    <span class="n">adata_ga</span><span class="p">,</span>
    <span class="n">rna_name_col</span><span class="o">=</span><span class="s2">&quot;gene_name&quot;</span><span class="p">,</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">coembed</span> <span class="o">=</span> <span class="n">build_aligned_coembedding</span><span class="p">(</span><span class="n">adata_gex</span><span class="p">,</span> <span class="n">adata_ga</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">rep_key</span> <span class="o">=</span> <span class="n">coembed</span><span class="o">.</span><span class="n">embedding_key</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[HVG] Joint (shared) HVGs: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coembed</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span><span class="si">:</span><span class="s2">,</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Transfer labels into ATAC via the shared PCA space and assemble the joint embedding:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">label_transfer_shared_pca</span><span class="p">(</span>
    <span class="n">adata_gex</span><span class="p">,</span>
    <span class="n">adata_ga</span><span class="p">,</span>
    <span class="n">label_key</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_pca_shared_aligned&quot;</span><span class="p">,</span>
    <span class="n">min_conf</span><span class="o">=</span><span class="mf">0.55</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">true_labels</span> <span class="o">=</span> <span class="n">adata_ga</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cell_type_original&quot;</span><span class="p">,</span> <span class="n">adata_ga</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cell_type&quot;</span><span class="p">))</span>
<span class="n">pred_labels</span> <span class="o">=</span> <span class="n">adata_ga</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;pred_cell_type&quot;</span><span class="p">]</span>
<span class="n">pred_labels_str</span> <span class="o">=</span> <span class="n">pred_labels</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">true_labels</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="o">~</span><span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">pred_labels</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pred_labels_str</span> <span class="o">!=</span> <span class="s2">&quot;unknown&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">mask</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
    <span class="n">nmi_score</span> <span class="o">=</span> <span class="n">normalized_mutual_info_score</span><span class="p">(</span>
        <span class="n">true_labels</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">),</span>
        <span class="n">pred_labels_str</span><span class="p">[</span><span class="n">mask</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[Metrics] NMI(pred_cell_type vs cell_type) = </span><span class="si">{</span><span class="n">nmi_score</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">joint_rep_key</span> <span class="o">=</span> <span class="n">rep_key</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">atac_prep</span><span class="o">.</span><span class="n">assemble_joint_embedding</span><span class="p">(</span><span class="n">joint_rep_key</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;RNA&quot;</span><span class="p">:</span> <span class="n">adata_gex</span><span class="p">})</span>
</pre></div>
</div>
<p>Plot the aligned embedding to inspect label transfer quality and clustering:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="n">joint_rep_key</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">joint_rep_key</span><span class="si">}</span><span class="s2">_leiden&quot;</span><span class="p">)</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;X_umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;X_umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">joint_rep_key</span><span class="si">}</span><span class="s2">_leiden&quot;</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
    <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Haihui Zhang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>